### PR TITLE
Add Poisson interval function for Mixnet

### PR DESCRIPTION
**Pull request by al8n**
Port [fisheryates.py](https://github.com/logos-co/nomos-specs/blob/master/mixnet/fisheryates.py) and [poisson.py](https://github.com/logos-co/nomos-specs/blob/master/mixnet/poisson.py) to Rust.

### CODE REVIEW

File: Cargo.toml
Comment (youngjoon-lee) [inline Cargo.toml:30]: How about having a `mixnet` crate on the root? because we’re gonna implement a lot of mixnet-specific logic soon. At my latest PR, I’ve deleted the previous `mixnet` crate, so you can recreate from scratch

File: mixnet-utils/src/lib.rs
Comment (youngjoon-lee) [inline mixnet-utils/src/lib.rs:23]: We may be able to reuse the existing fisheryates in Carnot https://github.com/logos-co/nomos-node/blob/3f2905817a92f0e8714b273183b2c0515004ea7f/consensus/carnot-engine/src/overlay/membership.rs#L16

File: Cargo.toml
Comment (youngjoon-lee) [reply Cargo.toml:30]: Let me merge my PRs now

File: mixnet/Cargo.toml
Comment (youngjoon-lee) [inline mixnet/Cargo.toml:2]: Package name should be modified.

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [inline mixnet/src/utils.rs:20]: Looks good in general. Approving this with some suggestions.

Please keep in mind that this function is not really necessary for logic actually for now. In the executable spec, this function is used only for testing (we may be able to refactor the executable spec, but anyway). In the future, we may need to use this function to dynamically update several configurations such as lambda or mu. So, I'm okay to keep this function here.

Also, it would be helpful to add a comment in the first method, that why we use exponential distribution for intervals, even though the function name says 'poisson'. It's explained in the spec as a comment. 

Lastly, it would be great to have a testcase to guarantee that the method you implemented work as intended, even though this function is very simple, because we may have regressions in the future.

**Review by youngjoon-lee — APPROVED** [review]

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [reply mixnet/src/utils.rs:20]: Let's not merge this now until #570 is merged.

File: mixnet/src/utils.rs
Comment (danielSanchezQ) [inline mixnet/src/utils.rs:14]: I would say we should abstract the randomness source. For two reasons:
* We may need to have tests that we can use deterministically (hence from a custom seed).
* We may have to use a custom random source for other reasons

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [reply mixnet/src/utils.rs:14]: Yes. It's necessary for testing for sure. The second reason is not really required for mixnet v1 spec, but still it would be worth.

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [inline mixnet/src/utils.rs:22]: > Also, it would be helpful to add a comment in the first method, that why we use exponential distribution for intervals, even though the function name says 'poisson'. It's explained in the spec as a comment.

What I meant in https://github.com/logos-co/nomos-node/pull/575#discussion_r1472388701 is that we need to mention why we use "exponential" distribution even though the name of this struct starts with `Poisson`. Someone may don't understand it if there is no clear explanation. It would be enough to just copy and paste the explanation in [nomos-specs](https://github.com/logos-co/nomos-specs/blob/9a54d90d1411de99647373914450bcc57520b92b/mixnet/poisson.py#L5)

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [inline mixnet/src/utils.rs:105]: Shouldn't we need to call the `PoissonExpInterval.interval(rate_per_min)` to test our code?

Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: i fixed the test: c71d53c1a27e2809e66bf360c5ea237528174cc9

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: I added comments: e6cdc9fb673cb73e87e949c65557efd65a1fb9a4. Also, I renamed the `PoissonExpInterval` struct to `Poisson`, because the struct has the `interval` function and we don't need to expose its implementation detail (the fact of using exp distribution for intervals) by its name.

File: mixnet/src/utils.rs
Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: @danielSanchezQ Could you please check the commit that Al made [129ef80](https://github.com/logos-co/nomos-node/pull/575/commits/129ef801e05ef0e1545ca0806cd187d2f8ff7b25), so we can merge this? On top of his commits, I pushed additional commits to polish this PR.

**Review by youngjoon-lee — APPROVED** [review]

File: mixnet/src/utils.rs
Comment (al8n) [reply mixnet/src/utils.rs]: Thanks! Sorry for the late reply. I do not know why I lost my GitHub notification from email.

Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: No worries. I was running this test on my local with this change, and I thought it's better to push my code to this PR instead of just leaving comments :) Please let me know if there is something to be changed in my commits.

File: mixnet/src/utils.rs
Comment (danielSanchezQ) [inline mixnet/src/utils.rs:25]: Why not generic function instead of a struct?

Comment (al8n) [reply mixnet/src/utils.rs:25]: To avoid always providing a `seed` to the function, I assume that we will invoke this function multiple times.

Comment (danielSanchezQ) [reply mixnet/src/utils.rs:25]: We may have to change the randomness source as well. @youngjoonlee can you confirm?

Comment (youngjoon-lee) [reply mixnet/src/utils.rs:25]: In our Mixnet specification, we don't strictly specify how randomness source has to be set for Poisson. 
Poisson distribution is used for both the interval of packet emissions (by msg senders) and the packet delays (by mix nodes). As long as all of them follow a Poisson distribution with the same "rate parameter" (aka. lambda) for a long time, the effect of mixing can be maintained as calculated in the [spec](https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#535f72f7b3eb4577adfde7f4124fde4b). From my understanding, the randomness source is not very important as long as the same rate parameter is used across the network. Implementers (including us) can decide on the strategy whether they'll change the source periodically. It's okay even if the same randomness source is shared across the network.

Similarly, Nym implementation uses `OsRng`: [reference1](https://github.com/nymtech/nym/blob/develop/common/client-core/src/client/cover_traffic_stream.rs#L109-L109) and [reference2](https://github.com/nymtech/nym/blob/develop/common/client-core/src/client/real_messages_control/mod.rs#L147-L147). (they're still [discussing](https://github.com/nymtech/nym/blob/3748ab77a132143d5fd1cd75dd06334d33294815/common/nymsphinx/src/utils/mod.rs#L11) which Rng is better, though).

So, in our implementation, I think we can go with the struct because we're not gonna change the Rng until we find any better strategy. 
(Nym implementation has a generic function instead :) But I think both is okay.)

Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: @al8n @danielSanchezQ This PR has stayed for a long time. Now, I'm taking over this PR. We should try to make PRs not stay opened for too long.

I just added a final commit 80e68d23, which just defines a generic function instead of struct. While implementing mixnet logic, I've found we actually don't need `seed`. Let's go with the simplest structure, as Nym does: https://github.com/nymtech/nym/blob/release/v1.1.22/common/nymsphinx/src/utils/mod.rs. I think we don't need to make things complicated before knowing what we really need. 

@danielSanchezQ Please turn off the `change required` sign in this PR, so I can merge this PR.

Comment (youngjoon-lee) [reply mixnet/src/utils.rs]: Also, renamed `utils.rs` to `poisson.rs`

**Review by danielSanchezQ — APPROVED** [review]

**Pull request merged**
